File: src/matlab/Lib/plot_segment_responses.m
%% plot_segment_responses.m
% 按通道绘制各推力等级的阶跃响应时序图
% 输入: data/csv/auv_data_merged.csv + auv_segments_final.csv
%
% 每个通道生成一张大图，包含:
%   - 各推力等级的速度/角速度响应叠加
%   - 对应指令输入
%   - 动力电源功率 (如有)

clear; close all;

% -------------------------------------------------------------------------
% 0. 路径 & 参数
% -------------------------------------------------------------------------
project_root = fileparts(fileparts(fileparts(fileparts(mfilename('fullpath')))));
if isempty(project_root)
    project_root = pwd;
end
data_dir = fullfile(project_root, 'data', 'csv');
out_dir  = fullfile(project_root, 'data', 'figures');
if ~exist(out_dir, 'dir'), mkdir(out_dir); end

% 读取主数据
fprintf('[1/4] 加载主数据...\n');
data_table = readtable(fullfile(data_dir, 'auv_data_merged.csv'));

% 读取段索引
fprintf('[2/4] 加载段索引...\n');
seg_table = readtable(fullfile(data_dir, 'auv_segments_final.csv'));

% 提取数组
t = data_table.stamp_time;
t0 = t(1);
motor_TX = data_table.motor_TX;
motor_TY = data_table.motor_TY;
motor_TZ = data_table.motor_TZ;
motor_MZ = data_table.motor_MZ;
linear_vx = data_table.linear_vx;
linear_vy = data_table.linear_vy;
angular_wz = data_table.angular_wz;
prop_pwr = data_table.prop_pwr_power;
ctrl_pwr = data_table.ctrl_pwr_power;
depth    = data_table.depth;
yaw      = data_table.yaw_rad;  % 注意列名可能不同

% -------------------------------------------------------------------------
% 3. 按通道分组绘制
% -------------------------------------------------------------------------
fprintf('[3/4] 绘图...\n');

channels = {'TX', 'TY', 'MZ'};
ch_labels = {'前进/后退 (Surge)', '右移/左移 (Sway)', '右转/左转 (Yaw)'};
cmd_cols  = {motor_TX, motor_TY, motor_MZ};
vel_cols  = {linear_vx, linear_vy, angular_wz};
vel_units = {'v_x (m/s)', 'v_y (m/s)', '\omega_z (°/s)'};

% 颜色映射: 20%->40%->60%->80%->100%
thrust_levels = [20, 40, 60, 80, 100];
colors = lines(length(thrust_levels));
level_color = containers.Map(thrust_levels, num2cell(colors, 2));

for ch_idx = 1:3
    ch_name = channels{ch_idx};
    cmd = cmd_cols{ch_idx};
    vel = vel_cols{ch_idx};
    v_unit = vel_units{ch_idx};

    % 筛选该通道的段
    ch_segs = seg_table(strcmp(seg_table.channel, ch_name), :);

    if isempty(ch_segs)
        fprintf('  %s: 无数据，跳过\n', ch_name);
        continue;
    end

    % 按推力等级和方向分组
    % 获取该通道所有出现的推力等级
    levels_in_data = unique(ch_segs.amplitude);
    n_levels = length(levels_in_data);

    % 分方向
    dir_pos = ch_segs(ch_segs.direction == 1, :);   % 前进/右移/右转
    dir_neg = ch_segs(ch_segs.direction == -1, :);  % 后退/左移/左转

    figure('Position', [50, 50, 1600, 900], 'Name', ch_name);

    % ---- 子图布局 ----
    % 左列: 正向 (前进/右移/右转)
    % 右列: 负向 (后退/左移/左转)
    % 行: 速度 + 指令 + 功率

    n_rows = 3;
    n_cols = 2;

    % ============ 正向 ============
    subplot_idx = @(r, c) (r-1)*n_cols + c;

    % -- 速度响应 --
    ax_vel_pos = subplot(n_rows, n_cols, subplot_idx(1, 1));
    hold on; grid on;
    title(sprintf('%s 正向', ch_labels{ch_idx}));

    % -- 指令 --
    ax_cmd_pos = subplot(n_rows, n_cols, subplot_idx(2, 1));
    hold on; grid on;
    ylabel('指令');

    % -- 功率 --
    ax_pwr_pos = subplot(n_rows, n_cols, subplot_idx(3, 1));
    hold on; grid on;
    xlabel('时间 (s)'); ylabel('动力功率 (W)');

    for i = 1:height(dir_pos)
        seg = dir_pos(i, :);
        idx_s = seg.idx_pad_start;
        idx_e = seg.idx_pad_end;
        seg_t = (t(idx_s:idx_e) - t(seg.idx_start))';  % 以段开始为零点
        seg_vel = vel(idx_s:idx_e);
        seg_cmd = cmd(idx_s:idx_e);
        seg_pwr = prop_pwr(idx_s:idx_e);
        amp_pct = seg.amp_pct{1};
        level_val = sscanf(amp_pct, '%d');
        clr = level_color(level_val);

        plot(ax_vel_pos, seg_t, seg_vel, 'Color', clr, 'LineWidth', 1.2);
        plot(ax_cmd_pos, seg_t, seg_cmd, 'Color', clr, 'LineWidth', 1.2);
        plot(ax_pwr_pos, seg_t, seg_pwr, 'Color', clr, 'LineWidth', 1);

        % 标注段区间
        t_start_seg = t(seg.idx_start) - t(seg.idx_start);
        t_end_seg = t(seg.idx_end) - t(seg.idx_start);
        xline(ax_vel_pos, t_start_seg, '--k', 'Alpha', 0.3);
        xline(ax_vel_pos, t_end_seg, '--k', 'Alpha', 0.3);
    end
    ylabel(ax_vel_pos, v_unit);

    % ============ 负向 ============
    ax_vel_neg = subplot(n_rows, n_cols, subplot_idx(1, 2));
    hold on; grid on;
    title(sprintf('%s 负向', ch_labels{ch_idx}));

    ax_cmd_neg = subplot(n_rows, n_cols, subplot_idx(2, 2));
    hold on; grid on;

    ax_pwr_neg = subplot(n_rows, n_cols, subplot_idx(3, 2));
    hold on; grid on;
    xlabel('时间 (s)');

    for i = 1:height(dir_neg)
        seg = dir_neg(i, :);
        idx_s = seg.idx_pad_start;
        idx_e = seg.idx_pad_end;
        seg_t = (t(idx_s:idx_e) - t(seg.idx_start))';
        seg_vel = vel(idx_s:idx_e);
        seg_cmd = cmd(idx_s:idx_e);
        seg_pwr = prop_pwr(idx_s:idx_e);
        amp_pct = seg.amp_pct{1};
        level_val = sscanf(amp_pct, '%d');
        clr = level_color(level_val);

        plot(ax_vel_neg, seg_t, seg_vel, 'Color', clr, 'LineWidth', 1.2);
        plot(ax_cmd_neg, seg_t, seg_cmd, 'Color', clr, 'LineWidth', 1.2);
        plot(ax_pwr_neg, seg_t, seg_pwr, 'Color', clr, 'LineWidth', 1);

        t_start_seg = t(seg.idx_start) - t(seg.idx_start);
        t_end_seg = t(seg.idx_end) - t(seg.idx_start);
        xline(ax_vel_neg, t_start_seg, '--k', 'Alpha', 0.3);
        xline(ax_vel_neg, t_end_seg, '--k', 'Alpha', 0.3);
    end

    % 图例
    L = {};
    for lv = thrust_levels
        L{end+1} = sprintf('%d%%', lv);
    end
    legend(ax_vel_pos, L, 'Location', 'best', 'FontSize', 8);
    legend(ax_vel_neg, L, 'Location', 'best', 'FontSize', 8);

    % Y 轴对齐
    linkaxes([ax_vel_pos, ax_vel_neg], 'y');
    linkaxes([ax_cmd_pos, ax_cmd_neg], 'y');

    % 保存
    sgtitle(sprintf('%s — 阶跃响应时序', ch_labels{ch_idx}), 'FontSize', 14);
    saveas(gcf, fullfile(out_dir, sprintf('response_%s.png', ch_name)));
    fprintf('  已保存: response_%s.png\n', ch_name);
end

% -------------------------------------------------------------------------
% 4. 汇总图: 稳态速度 vs 推力等级
% -------------------------------------------------------------------------
fprintf('[4/4] 稳态汇总图...\n');

figure('Position', [100, 100, 1400, 500]);

for ch_idx = 1:3
    ch_name = channels{ch_idx};
    ch_segs = seg_table(strcmp(seg_table.channel, ch_name), :);
    if isempty(ch_segs), continue; end

    subplot(1, 3, ch_idx);
    hold on; grid on;

    % 按推力等级汇总稳态速度均值
    levels = unique(ch_segs.amplitude);
    for d = [-1, 1]  % 负向/正向
        dir_segs = ch_segs(ch_segs.direction == d, :);
        for lv = levels'
            lv_segs = dir_segs(dir_segs.amplitude == lv, :);
            if isempty(lv_segs), continue; end

            if ch_idx == 3
                sv = lv_segs.steady_wz;
            elseif ch_idx == 2
                sv = lv_segs.steady_vy;
            else
                sv = lv_segs.steady_vx;
            end

            if height(lv_segs) == 1
                scatter(lv/100*100, sv, 60, 'filled');
            else
                scatter(repmat(lv/100*100, height(lv_segs), 1), sv, 60, 'filled');
            end
        end
    end

    xlabel('推力等级 (%)'); ylabel(v_unit);
    title(ch_labels{ch_idx});
end

saveas(gcf, fullfile(out_dir, 'steady_state_summary.png'));
fprintf('  已保存: steady_state_summary.png\n');

fprintf('\n完成! 图片保存在: %s\n', out_dir);

File: src/matlab/Lib/plot_segment_responses.py
#!/usr/bin/env python
"""
按通道绘制各推力等级的阶跃响应时序图
- 每个通道一张大图: 正向/负向 各子图(速度+指令+功率)
- 汇总图: 稳态速度 vs 推力等级
"""

import csv
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
from matplotlib.ticker import MaxNLocator
import os

# ---------------------------------------------------------------------------
# 配置
# ---------------------------------------------------------------------------
PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
DATA_DIR = os.path.join(PROJECT_ROOT, "data", "csv")
OUT_DIR = os.path.join(PROJECT_ROOT, "data", "figures")
os.makedirs(OUT_DIR, exist_ok=True)

# 中文字体
plt.rcParams['font.sans-serif'] = ['SimHei', 'Microsoft YaHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

# ---------------------------------------------------------------------------
# 1. 加载数据
# ---------------------------------------------------------------------------
print("[1/3] 加载数据...")

# 主数据
with open(os.path.join(DATA_DIR, "auv_data_merged.csv"), "r", encoding="utf-8") as f:
    reader = csv.DictReader(f)
    raw = {c: [] for c in reader.fieldnames}
    for row in reader:
        for c in reader.fieldnames:
            raw[c].append(row[c])

t = np.array([float(v) for v in raw["stamp_time"]])
motor_TX = np.array([int(float(v)) for v in raw["motor_TX"]])
motor_TY = np.array([int(float(v)) for v in raw["motor_TY"]])
motor_TZ = np.array([int(float(v)) for v in raw["motor_TZ"]])
motor_MZ = np.array([int(float(v)) for v in raw["motor_MZ"]])
linear_vx = np.array([float(v) for v in raw["linear_vx"]])
linear_vy = np.array([float(v) for v in raw["linear_vy"]])
angular_wz = np.array([float(v) for v in raw["angular_wz"]])
prop_pwr = np.array([float(v) for v in raw["prop_pwr_power"]])
depth = np.array([float(v) for v in raw["depth"]])
yaw_arr = np.array([float(v) for v in raw["yaw"]])

# 段索引
with open(os.path.join(DATA_DIR, "auv_segments_final.csv"), "r", encoding="utf-8") as f:
    segs = list(csv.DictReader(f))

print(f"  数据点: {len(t)}, 段数: {len(segs)}")

# ---------------------------------------------------------------------------
# 2. 绘图配置
# ---------------------------------------------------------------------------
print("[2/3] 绘制通道响应图...")

channels = [
    {"name": "TX", "label": "前进/后退 (Surge)", "cmd": motor_TX, "vel": linear_vx,
     "vlabel": "v_x (m/s)", "pos_name": "前进", "neg_name": "后退"},
    {"name": "TY", "label": "右移/左移 (Sway)", "cmd": motor_TY, "vel": linear_vy,
     "vlabel": "v_y (m/s)", "pos_name": "右移", "neg_name": "左移"},
    {"name": "MZ", "label": "右转/左转 (Yaw)", "cmd": motor_MZ, "vel": angular_wz,
     "vlabel": "ω_z (°/s)", "pos_name": "右转", "neg_name": "左转"},
]

THRUST_LEVELS = [20, 40, 60, 80, 100]
COLORS = plt.cm.viridis(np.linspace(0.1, 0.9, len(THRUST_LEVELS)))
LEVEL_COLOR = dict(zip(THRUST_LEVELS, COLORS))

for ch in channels:
    ch_name = ch["name"]
    ch_segs = [s for s in segs if s["channel"] == ch_name]
    if not ch_segs:
        print(f"  {ch_name}: 无数据，跳过")
        continue

    pos_segs = [s for s in ch_segs if int(s["direction"]) == 1]
    neg_segs = [s for s in ch_segs if int(s["direction"]) == -1]

    fig, axes = plt.subplots(3, 2, figsize=(18, 12), sharex='col')
    fig.suptitle(f'{ch["label"]} — 阶跃响应时序', fontsize=14, fontweight='bold')

    # ---- 正向 ----
    for i, seg in enumerate(pos_segs):
        idx_s = int(seg["idx_pad_start"])
        idx_e = int(seg["idx_pad_end"])
        seg_t = t[idx_s:idx_e] - t[int(seg["idx_start"])]  # 相对时间
        seg_vel = ch["vel"][idx_s:idx_e]
        seg_cmd = ch["cmd"][idx_s:idx_e]
        seg_pwr = prop_pwr[idx_s:idx_e]
        level = int(float(seg["amplitude"])) // 100  # 20, 40, ...
        clr = LEVEL_COLOR.get(level, 'gray')

        axes[0, 0].plot(seg_t, seg_vel, color=clr, linewidth=1.2)
        axes[1, 0].plot(seg_t, seg_cmd, color=clr, linewidth=1.2)
        axes[2, 0].plot(seg_t, seg_pwr, color=clr, linewidth=1.0)

        # 标出段区间
        t_ss = t[int(seg["idx_start"])] - t[int(seg["idx_start"])]
        t_se = t[int(seg["idx_end"])] - t[int(seg["idx_start"])]
        for ax_row in [0, 1, 2]:
            axes[ax_row, 0].axvline(t_ss, color='k', linestyle='--', alpha=0.25, linewidth=0.8)
            axes[ax_row, 0].axvline(t_se, color='k', linestyle='--', alpha=0.25, linewidth=0.8)

    axes[0, 0].set_title(ch["pos_name"], fontsize=12)
    axes[0, 0].set_ylabel(ch["vlabel"])
    axes[0, 0].grid(True, alpha=0.3)
    axes[1, 0].set_ylabel('指令')
    axes[1, 0].grid(True, alpha=0.3)
    axes[2, 0].set_ylabel('动力功率 (W)')
    axes[2, 0].set_xlabel('时间 (s)')
    axes[2, 0].grid(True, alpha=0.3)

    # ---- 负向 ----
    for i, seg in enumerate(neg_segs):
        idx_s = int(seg["idx_pad_start"])
        idx_e = int(seg["idx_pad_end"])
        seg_t = t[idx_s:idx_e] - t[int(seg["idx_start"])]
        seg_vel = ch["vel"][idx_s:idx_e]
        seg_cmd = ch["cmd"][idx_s:idx_e]
        seg_pwr = prop_pwr[idx_s:idx_e]
        level = int(float(seg["amplitude"])) // 100
        clr = LEVEL_COLOR.get(level, 'gray')

        axes[0, 1].plot(seg_t, seg_vel, color=clr, linewidth=1.2)
        axes[1, 1].plot(seg_t, seg_cmd, color=clr, linewidth=1.2)
        axes[2, 1].plot(seg_t, seg_pwr, color=clr, linewidth=1.0)

        t_ss = t[int(seg["idx_start"])] - t[int(seg["idx_start"])]
        t_se = t[int(seg["idx_end"])] - t[int(seg["idx_start"])]
        for ax_row in [0, 1, 2]:
            axes[ax_row, 1].axvline(t_ss, color='k', linestyle='--', alpha=0.25, linewidth=0.8)
            axes[ax_row, 1].axvline(t_se, color='k', linestyle='--', alpha=0.25, linewidth=0.8)

    axes[0, 1].set_title(ch["neg_name"], fontsize=12)
    axes[0, 1].grid(True, alpha=0.3)
    axes[1, 1].grid(True, alpha=0.3)
    axes[2, 1].set_xlabel('时间 (s)')
    axes[2, 1].grid(True, alpha=0.3)

    # Y 轴对齐
    axes[0, 0].sharey(axes[0, 1])
    axes[1, 0].sharey(axes[1, 1])

    # 图例
    legend_labels = [f'{lv}%' for lv in THRUST_LEVELS]
    axes[0, 0].legend(legend_labels, loc='best', fontsize=8, ncol=5)

    plt.tight_layout()
    out_path = os.path.join(OUT_DIR, f'response_{ch_name}.png')
    fig.savefig(out_path, dpi=150, bbox_inches='tight')
    plt.close(fig)
    print(f"  已保存: response_{ch_name}.png")

# ---------------------------------------------------------------------------
# 3. 稳态汇总图
# ---------------------------------------------------------------------------
print("[3/3] 稳态汇总...")

fig, axes = plt.subplots(1, 3, figsize=(16, 5))
fig.suptitle('稳态速度 vs 推力等级', fontsize=14, fontweight='bold')

for ch_idx, ch in enumerate(channels):
    ax = axes[ch_idx]
    ch_segs = [s for s in segs if s["channel"] == ch["name"]]
    if not ch_segs:
        continue

    for d, marker, label_suffix in [(-1, 'v', '(-)'), (1, '^', '(+)')]:
        dir_segs = [s for s in ch_segs if int(s["direction"]) == d]
        if not dir_segs:
            continue

        levels = sorted(set(int(float(s["amplitude"])) // 100 for s in dir_segs))
        for lv in levels:
            lv_segs = [s for s in dir_segs if int(float(s["amplitude"])) // 100 == lv]
            if ch_idx == 0:
                vals = [float(s["steady_vx"]) for s in lv_segs]
            elif ch_idx == 1:
                vals = [float(s["steady_vy"]) for s in lv_segs]
            else:
                vals = [float(s["steady_wz"]) for s in lv_segs]

            # 标注功率有效性
            has_pwr = [float(s["pp_seg_mean"]) > 30 for s in lv_segs]
            for j, (v, hp) in enumerate(zip(vals, has_pwr)):
                edge_clr = 'green' if hp else 'gray'
                ax.scatter(lv, abs(v), marker=marker, s=80,
                          facecolors='none' if not hp else COLORS[THRUST_LEVELS.index(lv)],
                          edgecolors=edge_clr, linewidths=1.5, zorder=3)

    ax.set_xlabel('推力等级 (%)')
    ax.set_ylabel(ch["vlabel"])
    ax.set_title(ch["label"])
    ax.grid(True, alpha=0.3)
    ax.set_xticks(THRUST_LEVELS)

# 图例
from matplotlib.lines import Line2D
legend_elements = [
    Line2D([0], [0], marker='^', color='w', markerfacecolor='gray', markersize=10, label='正向'),
    Line2D([0], [0], marker='v', color='w', markerfacecolor='gray', markersize=10, label='负向'),
    Line2D([0], [0], marker='o', color='w', markerfacecolor='gray', markersize=10,
           markeredgecolor='green', markeredgewidth=1.5, label='功率有效'),
    Line2D([0], [0], marker='o', color='w', markerfacecolor='gray', markersize=10,
           markeredgecolor='gray', markeredgewidth=1.5, label='功率无效'),
]
fig.legend(handles=legend_elements, loc='lower center', ncol=4, fontsize=9)

plt.tight_layout(rect=[0, 0.06, 1, 1])
out_path = os.path.join(OUT_DIR, 'steady_state_summary.png')
fig.savefig(out_path, dpi=150, bbox_inches='tight')
plt.close(fig)
print(f"  已保存: steady_state_summary.png")

print(f"\n完成! 图片保存在: {OUT_DIR}")
